### PR TITLE
Add append api and use it to deposit events

### PR DESCRIPTION
**Pull request by NikVolf**
polkadot companion: https://github.com/paritytech/polkadot/pull/1041

Either this or https://github.com/paritytech/substrate/pull/5740 (choosing one variant currently) is release critical @gnunicorn

### CODE REVIEW

**Comment by gavofyork** [comment]
@NikVolf doesn't build?

**Comment by NikVolf** [comment]
> doesn't build?

should be now

**Comment by gavofyork** [comment]
needs a companion PR?

File: primitives/io/src/lib.rs [hunk 1]
```diff
@@ -117,6 +117,11 @@ pub trait Storage {
 		Externalities::clear_prefix(*self, prefix)
 	}
 
+	/// Append to storage item (assumes it is in "Vec" format).
+	fn append(&mut self, key: &[u8], value: Vec<u8>) {
+		self.storage_append(key.to_vec(), value).expect("TODO");
```
Comment (gavofyork) [inline primitives/io/src/lib.rs:122]: unproven panicker.

File: primitives/state-machine/src/basic.rs [hunk 2]
```diff
@@ -257,6 +257,18 @@ impl Externalities for BasicExternalities {
...
 
+	fn storage_append(
+		&mut self,
+		key: Vec<u8>,
+		value: Vec<u8>,
+	) -> Result<(), sp_externalities::Error> {
+		self.place_storage(
+			key.clone(),
+			Some(crate::ext::append_to_storage(self.storage(&key).unwrap_or_default(), value)?)
+		);
```
Comment (gavofyork) [inline primitives/state-machine/src/basic.rs:268]:
```suggestion
		let previous = self.storage(&key).unwrap_or_default();
		let new = crate::ext::append_to_storage(previous, value)?;
		self.place_storage(key, Some(new));
```

File: primitives/state-machine/src/ext.rs [hunk 3]
```diff
@@ -551,6 +571,72 @@ where
...
+fn storage_error(err: &'static str) -> sp_externalities::Error {
+	sp_externalities::Error::StorageUpdateFailed(err)
+}
+
+fn extract_length_data(data: &[u8]) -> Result<(u32, usize, usize), sp_externalities::Error> {
+	use codec::CompactLen;
+
+	let len = u32::from(
+		Compact::<u32>::decode(&mut &data[..])
+			.map_err(|_| storage_error("Incorrent updated item encoding"))?
```
Comment (gavofyork) [inline primitives/state-machine/src/ext.rs:583]:
```suggestion
			.map_err(|_| storage_error("Incorrect updated item encoding"))?
```

**Comment by NikVolf** [comment]
@gavofyork addressed all issues, polkadot pr is up!

**Comment by gavofyork** [comment]
bot merge

**Pull request merged**